include/seqan3/range/view/async_buffer.hpp

```diff
@@ -0,0 +1,538 @@
+// View shortcut for functor.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::view
+{
+/*!\name General purpose views
+ * \{
+ */
+
+/*!\brief A view adapter that returns a concurrent-queue-like view over the underlying range.
+ * \tparam urng_t         The type of the range being processed. See below for requirements.
+ * \param[in,out] urange  The range being processed.
+ * \param[in] buffer_size Size of the buffer. Choose the size depending on the expected work per element.
+ * \returns A view that pre-fetches elements from the underlying range and provides a thread-safe interface.
+ *          See below for the properties of the returned range.
+ * \ingroup view
+ *
+ * \details
+ *
+ * **Header**
+ * ```cpp
+ *      #include <seqan3/range/view/async_buffer.hpp>
+ * ```
+ *
+ * ### Summary
+ *
+ * This view spawns a background thread that pre-fetches elements from the underlying range and stores them in a
+ * concurrent queue. Iterating over this view then pops elements out of the queue and returns them by value.
+ * This is primarily useful if dereferencing/incrementing the iterator of the underlying range
+ * is expensive, e.g. with SeqAn files which lazily perform I/O.
+ *
+ * Another advantage of this view is that multiple iterators can be created that are safe to iterate individually,
+ * even from different threads, i.e. you can use multiple threads to iterate safely over a single-pass input view
+ * with the added benefit of background pre-fetching.
+ *
+ * In technical terms: this view facilitates a single-producer, multi-consumer design; it's a range interface over
+ * a concurrent queue.
+ *
+ * ### Size of the buffer
+ *
+ * The `buffer_size` parameter should be chosen depending on the expected work per element, e.g. if the underlying
+ * range is an input file over short reads, a buffer size of 100 or 1000 could be beneficial; if on the other hand
+ * the file contains genome-sized sequences, it would be better to buffer only a single sequence (buffering 100
+ * sequences would result in the entire file being preloaded and likely consuming significant memory).
+ *
+ * ### Range consumption
+ *
+ * This view always moves elements from the underlying range into its buffer which means that the elements in
+ * the underlying range will be invalidated! For underlying ranges that are single-pass, this makes no difference, but
+ * it might be unexpected for multi-pass ranges (std::ranges::ForwardRange).
+ *
+ * Typically this adaptor is used when you want to consume the entire underlying range; there is no way to stop
+ * this view from buffering elements before the end of the underlying range is reached other than destructing it.
+ * If it is destructed before the end of the underlying range is reached, it will stop consuming elements, however,
+ * those elements that are currently in it's buffer are destructed with it.
```

Comment:
As my comment in the destructor suggests that would  be a data race. So after fixing this you might want to update the comment here as well.

Reply:
😱 No this is a very bad design. If you want to allow that consumers are not consuming everything you need a different signaling mechanism, i.e. consumers signal that they are done reading. The simplest way to do it is by using a barrier then. In that case you need to specify in the construction of the view how many consumers you are expecting and the destructor needs to wait for the consumers to signal they are done (arrived at the barrier managed by the view). Then you can safely destruct the queue even if it is not empy after closing since you are sure that no one is reading from it anymore. I need to think about how you can make it work nicely with the iterator.